File: 프로그래머스/2019 KAKAO BLIND RECRUITMENT/오픈채팅방.py
def solution(record):
    from collections import defaultdict
    chat = []
    name_list = defaultdict(str)
    for r in record:
            string = r.split()
            if string[0] != 'Leave':
                type,user_id,nickname = string[0],string[1],string[2]
                name_list[user_id] = nickname
            else:
                type,user_id = string[0],string[1]
                
            if type == 'Enter' or type == 'Leave':
                chat.append([user_id,type])
    answer = []
    for user_id,type in chat:
        if type == 'Enter':
            answer.append(name_list[user_id]+"님이 들어왔습니다.")
        else:
            answer.append(name_list[user_id]+"님이 나갔습니다.")
    return answer

File: 프로그래머스/lv2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        HashMap<String, String> nickName = new HashMap<>();
        HashMap<String, String> msg = new HashMap<>();
        msg.put("Enter", "님이 들어왔습니다.");
        msg.put("Leave", "님이 나갔습니다.");
        
        
        // o(n)
        for (String s : record) {
            String[] split = s.split(" ");
            if (split.length == 3) {
                nickName.put(split[1], split[2]);
            }
        }
        
        ArrayList<String> answer = new ArrayList<>();
        
        for (String s : record) {
            String[] split = s.split(" ");
            if (msg.containsKey(split[0])) {
                answer.add(nickName.get(split[1]) + msg.get(split[0]));
            }
        }
        
        
        return answer.toArray(new String[0]);
    }
}